sdks/python/apache_beam/io/filesystemio.py

```diff
@@ -0,0 +1,277 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Utilities for ``FileSystem`` implementations."""
+
+import abc
+import io
+import os
+
+__all__ = ['Downloader', 'Uploader', 'DownloaderStream', 'UploaderStream',
+           'PipeStream']
+
+
+class Downloader(object):
+  """Download interface for a single file.
+
+  Implementations should support random access reads.
+  """
+
+  __metaclass__ = abc.ABCMeta
+
+  @abc.abstractproperty
+  def size(self):
+    """Size of file to download."""
+
+  @abc.abstractmethod
+  def get_range(self, start, end):
+    """Retrieve a given byte range from this download, inclusive.
```

Comment:
<!--new_thread; commit:79815e6df1d49c271a6c44e37fb2a3c46be0aed5; resolved:0-->
Can we make the range half-open, i.e. [start, end)?  We used inclusive indices because we used the apitools library for GCS, and the reason the apitools library uses inclusive indices is because the HTTP range header uses inclusive indices (https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests), but that doesn't seem like a great reason for our interface to be inclusive too.